man/objhash-ng.1
.TH OBJHASH-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
objhash-ng: display information about ELF files.

.SH SYNOPSIS
objhash-ng [\fB\-H\fR|\fB\-\-help\fR]
           [\fB\-V\fR|\fB\-\-version\fR]
           [\fB\-s\fR|\fB\-\-sections\fR]
           [\fB\-h\fR|\fB\-\-headers\fR]
           [\fB\-a\fR|\fB\-\-all\fR]
           [\fB\-z\fR|\fB\-\-show\-size\fR]
           [\fB\-x\fR <number or name>|\fB\-\-hex\-dump=\fR<number or name>]
           [\fB\-p\fR <number or name>|\fB\-\-string\-dump=\fR<number or name>]
           [\fB\-P\fR <size>|\fB\-\-piecewise=\fR<size>]
           [\fB\-T\fR <size>|\fB\-\-threshold=\fR<size>]
           [\fB\-\-md5\fR]
           [\fB\-\-sha1\fR]
           [\fB\-\-sha256\fR]
           [\fB\-\-sha512\fR]
           [\fB\-\-ssdeep\fR]
           \fIelffile\fR...

.SH DESCRIPTION
\fBobjhash-ng\fR generates hashes of ELF format object files for malware detection and digitial forensics. The utility is capable of fuzz-hashing and context-triggered piecewise hashing. The options control what particular information to display.

\fIobjfile...\fR are the object files to be examined. 32-bit and 64-bit ELF files are supported.

.SH OPTIONS
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to objhash-ng and exit.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of objhash-ng and exit.
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD 0
.IP "\fB\-\-sections\fR" 4
.IX Item "--sections"
.PD
Displays the information contained in the file's section headers and report the SHA-256 for the section header.
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-headers\fR" 4
.IX Item "--headers"
.PD
Displays the information contained in the file's section headers and report the SHA-256 for the section data.
.IP "\fB\-a\fR" 4
.IX Item "-a"
.PD 0
.IP "\fB\-\-all\fR" 4
.IX Item "--all"
.PD
Equivalent to specifying: \fB\-\-sections\fR and \fB\-\-headers\fR.
.IP "\fB\-z\fR" 4
.IX Item "-z"
.PD 0
.IP "\fB\-\-show\-size\fR" 4
.IX Item "--show-size"
.PD
Display the file size before hash.
.IP "\fB\-x <number or name>\fR" 4
.IX Item "-x <number or name>"
.PD 0
.IP "\fB\-\-hex\-dump=<number or name>\fR" 4
.IX Item "--hex-dump=<number or name>"
.PD
Displays the contents of the indicated section as a hexadecimal bytes and report the SHA-256.
.IP "\fB\-p <number or name>\fR" 4
.IX Item "-p <number or name>"
.PD 0
.IP "\fB\-\-string\-dump=<number or name>\fR" 4
.IX Item "--string-dump=<number or name>"
.PD
Displays the contents of the indicated section as printable strings and report the SHA-256.
.IP "\fB\-P <size>\fR" 4
.IX Item "-P <size>"
.PD 0
.IP "\fB\-\-piecewise=<size>\fR" 4
.IX Item "--piecewise=<size>"
.PD
Piecewise mode. Breaks files into chunks before hashing.
.IP "\fB\-T <size>\fR" 4
.IX Item "-T <size>"
.PD 0
.IP "\fB\-\-threshold=<size>\fR" 4
.IX Item "--threshold=<size>"
.PD
Threshold mode. Only hash files smaller than the given the threshold.
.IP "\fB\-\-md5\fR" 4
.IX Item "--md5"
.PD 0
.IP "\fB\-\-sha1\fR" 4
.IX Item "--sha1"
.IP "\fB\-\-sha256\fR" 4
.IX Item "--sha256"
.IP "\fB\-\-sha512\fR" 4
.IX Item "--sha512"
.IP "\fB\-\-ssdeep\fR" 4
.IX Item "--ssdeep"
.PD
Encodes the input file using the specified cryptographic hash.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/objhash-ng.c
#include "buffer.h"
#include "printf.h"
#include "objhash.h"
#include "options.h"

#include "static/objhash-ng.ci"

#define THIS_NAME "objhash-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTIONELF);
  n += usage_synopsis0(o, name, args);
  n += usage_synopsis3(o, name, zOBJHASHARGS1);
  n += usage_objname(o, name, "elffile");
  n += printf_eol();
  n += usage_description(o, name, args, zOBJHASHDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options2(o, name, zOBJHASHARGS1, NULL, NULL);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_objhash(poptions_t o, int argc, char** argv, char* name) {
  if (argc < 1) {
    return usage(o, THIS_NAME, zOBJHASHARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 256);
      MALLOCA(char, arg1, 256);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, "--hex-dump")) {
          oinsertsname(o, ACT_HEXDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--string-dump")) {
          oinsertsname(o, ACT_STRDUMP8, arg1);
        } else if (0 == xstrcmp(arg0, "--piecewise")) {
          oinsertvalue(o, ACT_PIECEWISE, atovalue(arg1));
        } else if (0 == xstrcmp(arg0, "--threshold")) {
          oinsertvalue(o, ACT_THRESHOLD, atovalue(arg1));
        } else {
          return odeath(o, THIS_NAME, arg0 + 2);
        }
      } else {
        imode_t action = get_options2(o, zOBJHASHARGS, argv[i]);
        if (0 == action) {
          action = get_options2(o, zOBJHASHARGS1, argv[i]);
          if (0 == action) {
            return odeath(o, THIS_NAME, argv[i] + 2);
          }
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0]) {
      if (0 == xstrcmp(argv[i], "-x")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_HEXDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-p")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_STRDUMP8, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-P")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertvalue(o, ACT_PIECEWISE, atovalue(argv[++i]));
      } else if (0 == xstrcmp(argv[i], "-T")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertvalue(o, ACT_THRESHOLD, atovalue(argv[++i]));
      } else {
        imode_t action = get_options1(o, zOBJHASHARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (0 == o->inpname0[0]) {
      xstrncpy(o->inpname0, argv[i], NELEMENTS(o->inpname0));
    } else if (0 == o->inpname1[0]) {
      xstrncpy(o->inpname1, argv[i], NELEMENTS(o->inpname1));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  if (o->action & OPTPROGRAM_VERSION) {
    return version0(o, THIS_NAME, zOBJHASHARGS);
  }

  if (o->action & OPTPROGRAM_HELP) {
    return usage(o, THIS_NAME, zOBJHASHARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_objhash(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname[0]) {
      pbuffer_t p = bopen(o->inpname);
      if (p) {
        x = objhash(p, o);
      } else {
        printf_e("'%s': no such file.", o->inpname);
      }

      bfree(p);
    }

    ofree(o);
  }

  return x;
}

